### ISSUE TITLE
No event fired when mouse focus is lost

### ISSUE THREAD

**Issue by jmdejong** [fifengine/fifengine#1051]
The main panel (not sure how it's called) will fire mouse events on a listener you attach to it, for example mouseMoved events.
However, when moving the mouse into some widget, the mouseMoved events will not be fired anymore.
This wouldn't be a problem if the panel would be told that it stops getting mouse events, but there is none.

If a program keeps track of the mouse position by updating it on the mouseMoved events (which seems the most sensible solution) it will incorrectly think that the mouse is still at the last position before entering the widget.

A good solution would be either to fire a mouseExited event in this case, or to make a new event for this situation, for example mouseFocusLost.
With such an event a program can properly react to not getting mouse events anymore.

**Comment by helios2000** [comment]
Is that Fifechan related?
If so, please close the issue and open a new one for Fifechan. Did you start a new project? If this is the case, please use Cegui and not Fifechan. We want to get rid of Fifechan. It's only there because of Unknown Horizon.

I'm not sure what do you mean with 'main panel'. Fife have only an main window. If you do mean the main window then you do need a [CommandListener](https://github.com/fifengine/fifengine/blob/master/engine/core/eventchannel/eventchannel.i#L104). Here is the list with [command types](https://github.com/fifengine/fifengine/blob/master/engine/core/eventchannel/command/commandids.h#L41)
It works much like MouseListener.

**Comment by jmdejong** [comment]
I don't know if this is a Fifechan bug or a missing feature in the Fife design.

This is for unknown horizons, but it's probably good to look if the behaviour is the same with Cegui.
Is there documentation for installing/configuring Cegui for Fife?
I have tried installing Cegui from the AUR (https://aur.archlinux.org/packages/cegui/), but that didn't work for me.

With the main panel I mean the view of the isometric map (as opposed to all buttons and/or widgets in the overlay)

**Comment by helios2000** [comment]
Well, all that sounds pretty much like a Fifechan problem. If you want to know how event injection works, look here: https://github.com/fifengine/fifengine/blob/master/engine/core/gui/fifechan/fifechanmanager.cpp#L132
Events for Cegui are treated differently. So there should be no problem.

Besides from that, UH do very weird things with events and Fifechan and own Widgets, so it could also be an UH only problem.

@LinuxDonald should be able to help you with Cegui on linux.

**Comment by LinuxDonald** [comment]
last cegui release is very old and make maybe problems under linux.
I talked with him in the last hour about cegui compile.

**Comment by jmdejong** [comment]
So what's the expected behaviour? When the mouse moves over a button, then whatever was getting the mouseMoved events before should get a mouseExit?
I could try if I could fix that in fifechan

**Comment by LinuxDonald** [comment]
@helios2000 librocket is dead since 2016

**Comment by jmdejong** [comment]
@helios2000 In that case I suppose it's a fifechan bug indeed.
I'm going to try if I can fix it there.

Do you know how to refer to the main view of the map? And do you know what type it is? Is it a widget?

**Comment by helios2000** [comment]
The main view of the map don't help there. It's the window itself or even better the surface / context of it.
In case of Fifechan, on 'top' of it there is an invisible Fifechan Container (fcn_topcontainer) (C++). All other containers like VBox, HBox, Window are add to it on. Containers that are added to the TopContainer are called TopWidgets (Python).

**Comment by jmdejong** [comment]
I just found out that the mouseevents of the main view are not fired from fifechan.
Fifechan only fires mouse events on widgets.
Fife independently gets the mouse events from SDL in eventmanager.cpp

This is also where the check happens if the mouse event should be fired on the main view:
https://github.com/fifengine/fifengine/blob/41fe07971f58e693983520b466ecfa401f6cf37a/engine/core/eventchannel/eventmanager.cpp#L720

I'm not sure if this structure is a good idea, but I prefer not overhauling it, so the best thing is to fix it here.
This means that this is a bug in Fife, not Fifechan.

Tomorrow I'll try to see if I can work on it further.

**Comment by helios2000** [comment]
Well, then I still don't get your problem.
Events are pooled by SDL then send to EventManager and there to all different sorts of listeners. In case of Fifechan I posted already a link to the related code in FifechanManager.
The structure works, maybe there are better ways to receive the same behavior but thats not part of this issue.
Line 720 only checks if the event is marked as consumed by dispatchSdlEvent(e.g. FifechanManager) and if so it checks if an MouseFilter exists.

**Comment by jmdejong** [comment]
My problem is that the main view sometimes gets mouse move events and sometimes it doesn't (when the mouse is over a widget), and there is no way to detect when it will stop getting mouse move events.

I didn't mean exactly line 720. I meant to link to the code around. 723 may have been better to link to.

I want to replace:

    if (consumed) {
        return;
    }
    dispatchMouseEvent(mouseevt);

with:

    if (consumed) {
        if (!m_lastMouseEventConsumed){
            //dispatch mouseExited
        }
    } else {
        if (m_lastMouseEventConsumed){
            //dispatch mouseEntered
        }
        dispatchMouseEvent(mouseevt);
    }
    m_lastMouseEventConsumed = consumed;

I'll make a PR when I have figured out how to do what I wrote in comments (making a mouse event to dispatch).

**Comment by helios2000** [comment]
We are spinning in a circle. But I'm curious about the PR.

**Comment by jmdejong** [comment]
So, the core of the original problem is that the mouse stops firing mousemoved events to the main view when the mouse is hovering over a widget. When this happens there is no message, and no way to tell that it will stop firing mousemoved events.

I just realized that an alternative solution to this is to keep firing mousemoved events to the main view when the mouse is over a widget.
After some digging I found out that this is exactly what MouseFilters can do.
